src/master/heartbeat_cluster.hpp

```diff
@@ -0,0 +1,191 @@
+      ~node_entry () = default;
+
+      node_entry (const node_entry &other); // Copy c-tor
+      node_entry &operator= (const node_entry &other); // Copy assignment
+
+      enum node_state
+      {
+	UNKNOWN = 0,
+	SLAVE = 1,
+	TO_BE_MASTER = 2,
+	TO_BE_SLAVE = 3,
+	MASTER = 4,
+	REPLICA = 5,
+	MAX
+      };
+
+    public: // TODO CBRD-22864 members should be private
+      priority_type priority;
+      node_state state;
+      short score;
+      short heartbeat_gap;
+      timeval last_recv_hbtime; // last received heartbeat time
+  };
+
+  /* heartbeat ping host entries */
+  class ping_host : public hostname_node
+  {
+    public:
+      ping_host () = delete;
+      explicit ping_host (std::string hostname);
+      ~ping_host () = default;
+
+      void ping ();
+      bool is_ping_successful ();
+
+      enum ping_result
+      {
+	UNKNOWN = -1,
+	SUCCESS = 0,
+	USELESS_HOST = 1,
+	SYS_ERR = 2,
+	FAILURE = 3
+      };
+
+    public: // TODO CBRD-22864 members should be private
+      ping_result result;
+  };
+
+  /* heartbeat unidentified host entries */
+  class ui_node : public hostname_node
+  {
+    public:
+      explicit ui_node (std::string hostname, std::string group_id, const sockaddr_in &sockaddr, int v_result);
+      ~ui_node () = default;
+
+      void set_last_recv_time_to_now ();
+
+    public: // TODO CBRD-22864 members should be private
+      std::string group_id;
+      sockaddr_in saddr;
+      std::chrono::system_clock::time_point last_recv_time;
+      int v_result;
+  };
+
+  class cluster
+  {
+    public:
+      cluster ();
+
+      cluster (const cluster &other); // Copy c-tor
+      cluster &operator= (const cluster &other); // Copy assignment
+
+      ~cluster ();
+
+      int init ();
+      void destroy ();
+      int reload ();
+
+      int listen ();
+      void stop ();
+
+      node_entry *find_node (const std::string &node_hostname);
+
+      void remove_ui_node (ui_node *&node);
+      void cleanup_ui_nodes ();
+      ui_node *find_ui_node (const std::string &node_hostname, const std::string &node_group_id,
+			     const sockaddr_in &sockaddr) const;
+      ui_node *insert_ui_node (const std::string &node_hostname, const std::string &node_group_id,
+			       const sockaddr_in &sockaddr, const int v_result);
+
+      bool check_valid_ping_host ();
+
+    private:
+      int init_nodes ();
+      int init_replica_nodes ();
+      void init_ping_hosts ();
+
+    public: // TODO CBRD-22864 members should be private
+      pthread_mutex_t lock; // TODO CBRD-22864 replace with std::mutex
+
+      SOCKET sfd;
+
+      node_entry::node_state state;
+      std::string group_id;
+      std::string hostname;
+
+      std::list<node_entry *> nodes;
+
+      node_entry *myself;
+      node_entry *master;
+
+      bool shutdown;
+      bool hide_to_demote;
+      bool is_isolated;
+      bool is_ping_check_enabled;
+
+      std::list<ui_node *> ui_nodes;
```

Comment:
Wouldn't it be easier to manage std::list<node_entry>, std::list<ui_node>, std::list<ping_host> instead of lists of pointers?

Reply:
I changed ping_host list, for other lists is quite hard now to do this. I propose to get back to this issue once refactor for hb cluster is fully finished, we need a better encapsulation for cluster class